### PR TITLE
fix(snownet): replace `Allocation` if credentials to relay change

### PR DESCRIPTION
**Pull request by thomaseizinger**
When a relay restarts, its credentials change but the socket we use to connect to it might not. Because we upsert `Allocation`s within a `snownet::Node` based on the socket, such a change is currently not picked up.

Instead, we now check whether an existing allocation uses the same credentials and if it doesn't we throw the old one away and use the new one instead.

### CODE REVIEW

File: rust/connlib/snownet/src/node.rs
Comment (thomaseizinger) [inline rust/connlib/snownet/src/node.rs:848]: I started a discussion with the str0m maintainers on whether we should call `invalidate_candidate` here on all agents and perhaps trigger an ICE restart. it could be that we are just removing the nominated candidate pair and thus need to trigger an ICE restart but at present, `invalidate_candidate` doesn't tell me.

Discussion is here: https://str0m.zulipchat.com/#narrow/stream/377845-general/topic/invalidating-candidates

Comment (thomaseizinger) [reply rust/connlib/snownet/src/node.rs:848]: We don't have to block this PR on that though. If a relay restarts, the connections won't work anymore anyway so this is just an optimisation to perhaps rescue some of the connections through an ICE restart instead of having to fail the connection via the tunnel.

**Review by conectado — APPROVED** [review]

**Pull request merged**
